### ISSUE TITLE
ESP32_BLE_Arduino setting Manufacturor data is not 'zero' tolerant

### ISSUE THREAD

**Issue by iafilius** [nkolban/esp32-snippets#778]
Hi,

Working on Arduino, 
Installed latest development SDK (1.0.1-rc2). Bluetooth name=ESP32 BLE Arduino version=1.0.1 installed (library.properties)

While copying the BLE ManufacturerData from a real device to simulate the device.
I noticed the data is treated like a string, so a zero char half-way is treated like an end of string.
But in my case a zero is in the Manufacturor data, and can't set vendor_ID or additional vendor specific data properly  when it contains a zero with this code.

Minimalistic code to reproduce:
```
 BLEAdvertisementData scan_resp_advertisementData;
 //                               #ID-H   #ID-L data1  data2  data3  data4, zero
 const char manufacturerData[] = { 0xFF,  0x01 , 0x01,  0xFF,  0x00,  0x40 ,0x00};
 scan_resp_advertisementData.setManufacturerData(manufacturerData);
 ```
 In example this case data3 is zero, but the manufacturer ID bytes) may contain a zero as well, which behaves then not as desired. When looking with a BLE 'scanner' all after the first zero is basically chopped off.
 
 Desired:
 Change the code, to a byte array instead of string, and add an additional parameter <length> to the setManufacturerData API, to make the ManufacturerData 'zero' tolerant.
 
 Proposal for making a pull request and while keep using the std::string construction as used in multiple places.

 Current:
```
 /**
 * @brief Set manufacturer specific data.
 * @param [in] data Manufacturer data.
 */
--- void BLEAdvertisementData::setManufacturerData(std::string data) {
        ESP_LOGD("BLEAdvertisementData", ">> setManufacturerData");
        char cdata[2];
---        cdata[0] = data.length() + 1;
        cdata[1] = ESP_BLE_AD_MANUFACTURER_SPECIFIC_TYPE;  // 0xff
---     addData(std::string(cdata, 2) + data);
        ESP_LOGD("BLEAdvertisementData", "<< setManufacturerData");
} // setManufacturerData
```

 New (untested/proposal):
```
+++ void BLEAdvertisementData::setManufacturerData(std::string data, size_t len) {
        ESP_LOGD("BLEAdvertisementData", ">> setManufacturerData");
        char cdata[2];
+++        cdata[0] = len + 1;
        cdata[1] = ESP_BLE_AD_MANUFACTURER_SPECIFIC_TYPE;  // 0xff
+++     addData(std::string(cdata, 2) + std::string(data, len));
        ESP_LOGD("BLEAdvertisementData", "<< setManufacturerData");
} // setManufacturerData
``` 
 
 When this is the way to go, shall I make a pull request for it?

Regards,

Arjan

**Comment by iafilius** [comment]
Hi, an update after some additional research and debugging.

After some research I came to the conclusion the excisting code works perfectly well, but slightly intolerant to possible wrong usage.


non working minimalisting demo (data get's copped of at first zero):

```
 BLEAdvertisementData scan_resp_advertisementData;
 //                               #ID-H   #ID-L data1  data2  data3  data4, zero
 const char manufacturerData[] = { 0xFF,  0x00 , 0x01,  0xFF,  0x00,  0x40 ,0x00};
 scan_resp_advertisementData.setManufacturerData(manufacturerData);
```

Updated working demo:
```
 BLEAdvertisementData scan_resp_advertisementData;
 //                               #ID-H   #ID-L data1  data2  data3  data4, zero
 std::string manufacturerData = { 0xFF,  0x00 , 0x01,  0xFF,  0x00,  0x40 ,0x00};
 scan_resp_advertisementData.setManufacturerData(manufacturerData);
```

scan_resp_advertisementData.getPayload().length() returns now 9, which is perfectly fine (instead 3 of non working demo above)


So it is working as intended, but in case of using zero's in the string, usage of std::string type isn't optional!
As using zero's in the Manufacturer data is likely to happen, perhaps a recommendation is to enforce the std::string type instead silent type conversion from a char string[].


I leave the issue open (for a while), but as it works for me, there is no real reason to keep this open. So it may closed, or closed after implementing a more strict type checking.

Regards,


Arjan
